bytecomp/emitcode.ml

```diff
@@ -168,6 +168,37 @@ and slot_for_c_prim name =
 let events = ref ([] : debug_event list)
 let debug_dirs = ref String.Set.empty
 
+(* Map from absolute paths of .ml files to their sanitized paths. *)
+let sanitize_map = Hashtbl.create 17
+
+let sanitize_event ev =
+  let loc = ev.ev_loc in
+  let sloc = loc.loc_start in
+  let path = sloc.pos_fname in
+  let open Filename in
+  let unmapped_abspath =
+    if (is_relative path) then (concat (Sys.getcwd ()) path)
+    else path
+  in
+  let mapped_abspath = Location.rewrite_absolute_path unmapped_abspath in
+  (* If mapping did nothing, just keep the event as is. *)
+  if mapped_abspath = unmapped_abspath then ev
+  else begin
+    (* We want to share the same sanitized path to save space *)
+    let new_fname = match Hashtbl.find_opt sanitize_map mapped_abspath with
+    | Some saved_path -> saved_path
+    | None ->
+        Hashtbl.add sanitize_map mapped_abspath mapped_abspath;
+        mapped_abspath
```

Comment:
What is going on here is a bit subtle: you are not caching the computation of `Location.absolute_path` as one might expect, but hash-consing / maximal-sharing / interning the results of the computation after the fact.

Caching `Location.absolute_path` would be more natural to me, and it moves the code to a place with more users and where specific attention to this performance concern makes more sense (to me).

One subtlety is that, in your approach, the hash-consing table does *not* depend on the value of the BUILD_PATH_PREFIX_MAP variable, and would be robust to changes to this variable across calls to the function. (This is a realistic use-case with your debugger directive to set the variable.) If we changed it to caching the computation instead, then the cache would depend on the value. I see two ways to avoid any issue here:

1. Like you are doing here you could keep simply using a sharing/interning table.
2. Or we could store the last observed value of the BUILD_PATH_PREFIX_MAP variable, and clear the cache (before accessing it) if/when it has changed.

I wonder if we need caching/sharing at all, maybe the performance concern is moot. I am inclined to keep it in for the memory-saving benefits (there are many debug events, so deduplicating their location file names is not good).

Reply:
Thinking about this more: I am not sure where is the right place to have caching/sharing. We could also consider whole-event sharing: right now we don't have `sanitize_event ev == sanitize_event ev` whenever the paths inside get rewritten, which may also be a memory issue. (You did include a shortcut to ensure sharing in the case where the paths do not get rewritten, which is a nice first easy step, thanks.)

Reply:
I do not expect any events to be equal since they have a pc that is distinct for each event.